src/multiset.rs

```diff
@@ -0,0 +1,86 @@
+use rpds::RedBlackTreeMapSync;
+use std::hash::Hash;
+
+/// Immutable multiset implementation, which is threadsafe and hash stable, regardless of insertion order.
+///
+/// All methods that return a new multiset take ownership of the old multiset.
+#[derive(Debug, Hash, Eq, PartialEq, Clone)]
+pub(crate) struct MultiSet<T: Hash + Ord + Clone>(
+    /// All values should be > 0
+    RedBlackTreeMapSync<T, usize>,
+);
+
+impl<T: Hash + Ord + Clone> MultiSet<T> {
+    /// Create a new empty multiset.
+    pub(crate) fn new() -> Self {
+        MultiSet(RedBlackTreeMapSync::new_sync())
+    }
+
+    /// Check if the multiset contains a key.
+    pub(crate) fn contains(&self, value: &T) -> bool {
+        self.0.contains_key(value)
+    }
+
+    /// Return the total number of elements in the multiset.
+    pub(crate) fn len(&self) -> usize {
```

Comment:
I don't think this would work because we don't just want to count how many entries are in the map, but sum all the values. i.e. this length is will include duplicates.

Reply:
Oh, I see. Can we maintain the count as a struct field and assert that this code and that count are the same in debug mode? (Making len non-constant time is evil, worth slightly more complexity imo.)

Reply:
Sounds good, resolved.